DM-43101: Add exposure.can_see_sky metadata field

timj opened the pull request:
No attempt is made to generate a value when importing from an old universe. (replaces #972)

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [x] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

timj committed:
Add exposure.can_see_sky metadata field

timj committed:
Refresh pre-commit

timj committed:
Modernize ruff configuration in pyproject.toml

timj committed:
Use f-strings rather than .format()

timj committed:
Use collections.abc rather than typing

timj committed:
Use isinstance with |

timj committed:
Skip a test if this package is not pip installed

scons does not (yet) install standard python package metadata so
this test can fail if you have safir installed.

timj committed:
Use tuple rather than Tuple

timj committed:
Use | None and tuple rather than Optional[Tuple

timj committed:
Replace Union with TypeAlias and |

timj committed:
Replace Type with type[BaseModel]

timj committed:
Add UP check to ruff configuration

timj committed:
Archive new universe as v7

timj committed:
Add news fragment

timj commented on python/lsst/daf/butler/timespan_database_representation.py:
```diff
@@ -66,7 +66,7 @@ class _SpecialTimespanBound(enum.Enum):
 
 
-TimespanBound = Union[astropy.time.Time, _SpecialTimespanBound, None]
+TimespanBound: TypeAlias = astropy.time.Time | _SpecialTimespanBound | None
```
I'm not sure why this code is copied from `_timespan.py`.

TallJimbo replied:
It was just a mistake made when refactoring one module into two modules.  We should remove it and import instead it if it's actually used here.

timj commented on python/lsst/daf/butler/logging.py:
```diff
@@ -292,18 +292,18 @@ class ButlerLogRecords(_ButlerLogRecords):
             The records to seed this class with.
         """
-        return cls.model_construct(root=list(records))
+        return cast(Self, cls.model_construct(root=list(records)))
```
`cls.model_construct()` returns `ButlerLogRecords` and mypy can't work out that's not `Self`.

TallJimbo commented on python/lsst/daf/butler/json.py:
```diff
@@ -42,7 +44,7 @@ class SupportsSimple(Protocol):
     """
 
-    _serializedType: Type
+    _serializedType: type[BaseModel]
```
Should actually by `ClassVar[type[BaseModel]]`, I suspect.

Aside: this is correct for the code as it is now, but I think we really want to support things that are recognized by Pydantic other than `BaseModel` (like `typing.TypedDict`) as "simple" objects in the future.  I think that would involve making this into:

```
class SupportsSimple(Protocol, Generic[_T]):

    _adapter: ClassVar[pydantic.TypeAdapter[_T]]
```

timj replied:
Maybe, but `TypeAdapter` doesn't support `model_dump_json` which is what the code relies on (it's called `dump_json` with `TypeAdapter`).

I'll do your `ClassVar` fix.

TallJimbo reviewed (approved):
I'd prefer to revert the `Union -> |` commit and tell Ruff to either ignore that rule (ideal) or ignore at least the majority of those lines, as I think `Union` is better for anything multi-line (or potentially multi-line in the future).

I'm also not sold on the `Self` commit, as that is a subtle definition change rather than just notation, when inheritance is in play; I prefer not to use `Self` unless I want to say that derived classes must return the most-derived type.  Looks like we could be using `from __future__ import annotations` in `ButlerLogRecords` instead, though - probably dates from when pydantic couldn't handle that.

timj committed:
Remove quoting and union from butler log record annotations

timj committed:
Remove Union in simple type aliases

After discussion with Jim Bosch we have agreed that multi-line
type aliases can include Union for clarity and as a result of
that the UP007 warning is now ignored.

timj committed:
Remove time class that was unused due to copy and paste error

timj committed:
Fix description of v6 universe change

timj committed:
Fix Butler.get() type signature

This was missing from a previous ticket.
